recipes/cupy/meta.yaml

```diff
@@ -0,0 +1,59 @@
+{% set name = "cupy" %}
+{% set version = "6.4.0" %}
+{% set sha256 = "2a3cd1812f043ba9451b8ca4e5a4c31f9acc486c003f6970836e1a46c735c82e" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/cupy/cupy/archive/v{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [not (linux64 and cuda_compiler_version == "9.2")]
+  script: "{{ PYTHON }} -m pip install . -vv"
+  missing_dso_whitelist:
+    - "*/libcuda.*"
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ compiler("cuda") }}
+
+  host:
+    - python
+    - pip
+    - setuptools
+    - cython >=0.24.0
+    - cudnn
+    - fastrlock >=0.3
+    - nccl
+
+  run:
+    - python
+    - setuptools
+    - {{ pin_compatible("cudnn") }}
```

Comment:
Maybe I misunderstood: why just pinned `cudnn` but not `nccl`? There are incompatibilities even within the v2 series, and I am not sure if the built `cupy.cuda.nccl` module would be happy with a different nccl version discovered at runtime.

Reply:
FYI, the CuPy team bundles a specific NCCL version in the CuPy wheel.

Reply:
Yeah not pinning NCCL is dangerous. The NCCL version is determined at built time:
https://github.com/cupy/cupy/blob/b9db374f4d0ccee358a998d7b283248f77cef351/cupy/cuda/cupy_nccl.h#L131-L133

Reply:
I think the plan is to pin that globally https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/309#issuecomment-545685035

Reply:
Thanks, @CJ-Wright. I wasn't aware of that discussion. 

But still, don't we need a  `{{ pin_compatible("nccl") }}` here? Once a CuPy package is built, it's pinned to the NCCL version it was built against. I feel the global pin might be a bit fragile when a newer NCCL comes out.

Reply:
`nccl` itself [uses `run_exports`]( https://github.com/conda-forge/nccl-feedstock/blob/db95a970c6d4ed74873822ffd30e7c2fcff001a3/recipe/meta.yaml#L23-L25 ). So it doesn't need `pin_compatible`. The lower bound as CJ mentions should be handled at the global level. Though I've included a workaround in this recipe until we integrate that PR.

For `cudnn` it would be great if the package also used `run_exports`. I've raised issue ( https://github.com/AnacondaRecipes/aggregate/issues/173 ) on this point. Though it could be useful to also pin `cudnn` globally as a stopgap.

Reply:
Thanks for the explanation of `run_exports`, @jakirkham. But I am afraid the issue you referred to is mistaken: 
> The NCCL API only changed between NCCL 1.x and NCCL 2.0.

This is not true. Even within v2 there're API incompatibilities, see how CuPy handles one of the cases here: 
https://github.com/cupy/cupy/blob/f19be9ca7d4c887faeaf6601649ca53668d6fdac/cupy/cuda/cupy_nccl.h#L180-L187

We really need a precise pin (`x.x.x`).

Reply:
Also here: https://github.com/cupy/cupy/blob/f19be9ca7d4c887faeaf6601649ca53668d6fdac/cupy/cuda/cupy_nccl.h#L251-L255

Reply:
@leofang, could you please comment on issue ( https://github.com/NVIDIA/nccl/issues/218 )? AIUI the NCCL developers intend to maintain that level of compatibility. So if NCCL is breaking the API/ABI, that's a bug.

Reply:
Actually can you please clarify something for me, are they simply adding new functionality or they breaking the API/ABI? These would be handled a bit differently from the Conda perspective.

Reply:
Just to explain a bit more (and hopefully give context to my question above), the way `pin_compatible` and `run_exports` work is more-or-less the same from the versioning standpoint. Basically they set a minimum supported version, which is whatever version was used to build the package. Then they allow any version greater or equal to that minimum up to whatever the maximum compatible version would be.

So a worked example of this with `nccl` based on PR ( https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/309 ) would be as follows.

1. `cupy` is built with `nccl` 2.4.6.1
2. `cupy`'s minimum supported `nccl` version is now 2.4.6.1
3. `cupy`'s maximum supported `nccl` version is anything below 3 (so 2.x.x.x greater than or equal to 2.4.6.1).

If `nccl` only adds new features and does not change existing API/ABI in a breaking way (until 3), then the `cupy` package will continue to work with those newer versions. It simply will not use new APIs introduced in `nccl` without being rebuilt.

Does that make sense?

Reply:
It's also important that CuPy is built against at least the latest version of NCCL at the time of CuPy release so that CuPy doesn't define these empty functions and symbols clash.

Reply:
> So, as you said if 1 and 2 are satisfied, that's okay.

Is that correct @leofang? Want to make sure we get this right the first time. 🙂

Reply:
Yes I agree with @isuruf. I think @jakirkham your argument is valid as long as it's pinned at v2.4+. I was perhaps a bit over exaggerating when saying it was wrong without considering the possibility that new features can be added in an API/ABI compatible fashion. Thanks a lot for organizing the thoughts clearly and for hard work, @jakirkham. You have my 👍

Reply:
Note that https://github.com/conda-forge/staged-recipes/pull/9959#discussion_r338663623 is really important

If you are going with a minimum version of 2.4.6.1, then you should pin exactly or make sure there's no empty functions defined like that in CuPy with 2.4.6.1. (A workaround is to build with latest NCCL)

Reply:
Didn't realize that @leofang is a contributor. @leofang, do you mind making those empty functions `inline`?

Reply:
@isuruf, could you please open a CuPy issue about that?

Reply:
One things I can say, not sure if I answer @isuruf's questions:

For the empty functions that `return ncclSuccess;` there, they are either solely for the purpose of building Read the Docs and not used in real world (provided `CUPY_NO_CUDA` is defined), or they are skeletons that are guaranteed not called in runtime, as long as the pinned NCCL version matches (or is larger than, with the same major version) `NCCL_VERSION_CODE` set during build time. 

Pinning at 2.4.6.1 seems safe to me.

Reply:
That's true. Pinning at 2.4.6.1 exactly works too. But I think @jakirkham wanted to pin at 2.x.x.x>=2.4.6.1. To do that we must make sure CuPy doesn't export symbols like `ncclCommGetAsyncError` to prevent clashes. Making that function `inline` takes care of this issue. (I know that `ncclCommGetAsyncError` wouldn't get defined in CuPy if we build with 2.4.6.1, but that's just an example.)

Reply:
The other option would be to create wrapper functions with a different name that call NCCL or are stubbed out.

Reply:
True, but you'll need to make them `inline` too to avoid the overhead of calling the function.

Reply:
Yeah we probably want both. Compilers sometimes ignore `inline`, but adding it should at least let them know it is an option.

Reply:
I don't think we need to worry about symbol clash. In runtime if, say, `ncclCommGetAsyncError` is empty, it's guaranteed not called: https://github.com/cupy/cupy/blob/f19be9ca7d4c887faeaf6601649ca53668d6fdac/cupy/cuda/nccl.pyx#L429-L432. 
Moreover, if there's any clash, shouldn't the compiler error out or complain when building?

Reply:
I've raised issue ( https://github.com/cupy/cupy/issues/2568 ) on this point.

Reply:
Well the compiler might warn that the symbol is unused, but I don't think it will just get rid of the symbol altogether. There may be flags to enable this kind of elimination, but am pretty sure it is opt-in. I could be wrong about this though.

Reply:
Anyways we should probably move this discussion to CuPy issue ( https://github.com/cupy/cupy/issues/2568 ).

Reply:
Well currently we are using the latest major minor version of NCCL with CuPy. Now that we are aware of this issue, I think we just need to stay on top of it.

Reply:
Until that issue is resolved, shouldn't you pin exactly?

Reply:
The NCCL issue? We are using the latest major minor version of NCCL and the current release of CuPy does too. So there are not adding any symbols at the moment that will clash with NCCL. Also they've acknowledge the issue upstream and seem willing to work on it. We certainly need to keep an eye on things as new releases come out and adjust our strategy accordingly. That said, I think we are doing all we can with this issue ATM.